tuned/plugins/plugin_scheduler.py

```diff
@@ -273,28 +292,65 @@ def _tune_process_rt(self, pid, sched, prio):
 						% (pid, e))
 		return cont
 
-	def _store_orig_process_affinity(self, pid, affinity):
+	def _is_cgroup_affinity(self, affinity):
+		return str(affinity)[:7] == "cgroup."
+
+	def _store_orig_process_affinity(self, pid, affinity, is_cgroup = False):
 		try:
 			params = self._scheduler_original[pid]
 		except KeyError:
 			params = SchedulerParams(self._cmd)
 			self._scheduler_original[pid] = params
-		if params.affinity is None:
-			params.affinity = affinity
+		if params.affinity is None and params.cgroup is None:
+			if is_cgroup:
+				params.cgroup = affinity
+			else:
+				params.affinity = affinity
+
+	def _get_cgroup_affinity(self, pid):
+		# we cannot use procfs, because it uses comma ',' delimiter which
+		# can be ambiguous
+		for l in self._cmd.read_file("%s/%s/%s" % (consts.PROCFS_MOUNT_POINT, str(pid), "cgroup"), no_error = True).split("\n"):
+			try:
+				cgroup = l.split(":cpuset:")[1][1:]
+				return cgroup if cgroup != "" else "/"
+			except IndexError:
+				pass
+		return "/"
+
+	# it can be arbitrary cgroup even cgroup we didn't set, but it needs to be
+	# under "cgroup_mount_point"
+	def _set_cgroup(self, pid, cgroup):
+		cgroup = self._sanitize_cgroup_path(cgroup)
+		path = self._cgroup_mount_point
+		if cgroup != "/":
+			path = "%s/%s" % (path, cgroup)
+		self._cmd.write_to_file("%s/tasks" % path, str(pid), no_error = True)
+
+	def _parse_cgroup_affinity(self, cgroup):
+		# "cgroup.CGROUP"
+		cgroup = cgroup[7:]
+		is_cgroup = len(cgroup) > 0
+		return is_cgroup, cgroup
 
 	def _tune_process_affinity(self, pid, affinity, intersect = False):
 		cont = True
 		if affinity is None:
 			return cont
 		try:
-			prev_affinity = self._get_affinity(pid)
-			if intersect:
-				affinity = self._get_intersect_affinity(
-						prev_affinity, affinity,
-						affinity)
-			self._set_affinity(pid, affinity)
+			(is_cgroup, cgroup) = self._parse_cgroup_affinity(affinity)
```

Comment:
If I'm reading the code right, then `affinity` can be a list of CPUs. And if that list is has more than 7 elements, then this will return a false positive.

Reply:
Good catch, hopefully fixed.